Fix `/kill` behaving differently to `/minecraft:kill`

abulujayn opened the pull request:
### Information

This PR fixes #5842. 

### Details

**Proposed fix:**    

- Use Damageable#damage(double) instead of `Damageable#setHealth(double)`
- Use Float.MAX_VALUE as that is what `/minecraft:kill` uses


**Environments tested:**    

OS: Windows 11

Java version:  Corretto-21.0.3.9.1

- [x] Most recent Paper version (1.20.6-147-ver/1.20.6@e41d44f)
- [x] CraftBukkit/Spigot/Paper 1.12.2 (Just tested successful killing, not trigger of advancement)
- [x] CraftBukkit 1.8.8 (Just tested successful killing, not trigger of advancement)


**Demonstration:**    
Triggers `minecraft:entity_hurt_player` which was previously not being triggered as mentioned in #5842
![image](https://github.com/EssentialsX/Essentials/assets/13144755/0af1b38d-2bbd-4b0d-96d2-8ae5719269e8)

abulujayn committed:
Use `Damageable#damage(double)` instead of `Damageable#setHealth(double)`

Using Float.MAX_VALUE as that is what `/minecraft:kill` uses

Fixes #5842

abulujayn committed:
Merge branch '2.x' into fix_kill_command

CyberFlameGO commented:
I thought it intentionally was different to vanilla behavior - this change would also break /kill bypassing god mode iirc

abulujayn commented:
Thanks for pointing that out, it does indeed break when in god mode. If older versions had a damage cause, could just use that but doesn't seem they do. Other solution I can think of is to check for `Float.MAX_VALUE` since I can't think of any example of something other than `/kill` using that value.

abulujayn committed:
Fix god mode blocking `/kill` from working

- Remove phantom EntityDamageEvent
- Check if `/kill` command used by checking if `getDamage() == Float.MAX_VALUE`

abulujayn committed:
Merge branch '2.x' into fix_kill_command

GrantGryczan commented:
I haven't looked closely at this, but if your fix is to make `/kill` behave like `/minecraft:kill`, note that the original issue includes `/minecraft:kill` as being affected too. That is, if EssentialsX is installed, `/minecraft:kill` also doesn't trigger `entity_hurt_player`.

GrantGryczan commented:
Does `/kill` still work on creative and spectator players using this?

GrantGryczan commented:
Unlike `/kill` in vanilla, I just found that this makes `/kill` no longer work on creative players. The source damage type should be `minecraft:generic_kill` (not `minecraft:generic`, Spigot's default) in order to affect creative players. Unfortunately Spigot's API started letting you set the damage source only recently (I believe as of a Spigot version for MC 1.20.4, but not the first so this fix might have to only apply to 1.20.5+ to be completely safe), but I think it's fine if #5842 is only fixed for newer Minecraft versions (1.20.5+).

Here's how another plugin implemented this fix: https://github.com/ChanceSD/PvPManager/commit/aa358e9a55f9f4cebb20a3f52ee6b036f9e5ceab

In general, the condition vanilla Minecraft uses to check if an invulnerable player should be damaged is whether the source damage type is in the [`minecraft:bypasses_invulnerability` damage type tag](https://github.com/misode/mcmeta/blob/data/data/minecraft/tags/damage_type/bypasses_invulnerability.json) (which includes `minecraft:generic_kill` by default). EssentialsX god mode should check that same condition--it's another data-pack-breaking bug that it doesn't already, since data packs can create their own damage types and add them to the `minecraft:bypasses_invulnerability` tag.

Might want to apply this to `/suicide` too. `minecraft:out_of_world` is also a damage type that bypasses invulnerability if you want to use that instead for the sake of its death message. And if you don't want to use a vanilla damage type, then any custom instant-kill damage types should be added to the appropriate [damage type tags](https://github.com/misode/mcmeta/blob/data/data/minecraft/tags/damage_type) such as `minecraft:bypasses_invulnerability` (look at all of the damage type tags that `minecraft:generic_kill` is a part of).

abulujayn committed:
Merge branch 'EssentialsX:2.x' into fix_kill_command

GrantGryczan commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandkill.java:
```diff
@@ -29,12 +28,7 @@ protected void updatePlayer(final Server server, final CommandSource sender, fin
         if (sender.isPlayer() && user.isAuthorized("essentials.kill.exempt") && !ess.getUser(sender.getPlayer()).isAuthorized("essentials.kill.force")) {
             throw new PlayerExemptException("killExempt", matchPlayer.getDisplayName());
         }
-        final EntityDamageEvent ede = ess.getDamageEventProvider().callDamageEvent(matchPlayer, sender.isPlayer() && sender.getPlayer().getName().equals(matchPlayer.getName()) ? EntityDamageEvent.DamageCause.SUICIDE : EntityDamageEvent.DamageCause.CUSTOM, Float.MAX_VALUE);
```
Removing the damage sources here and all the cancellation and permission checks below definitely isn't the way to go. Please see [my previous comment on the PR](https://github.com/EssentialsX/Essentials/pull/5850#issuecomment-2223132714).

abulujayn commented:
Thanks for sharing! I did miss those things. I wonder how legacy versions of MC implemented the vanilla kill command to account for invulnerability 🤔. Might take a look at that. I'm sorry I haven't addressed this earlier, have been absolutely battered at work recently.

abulujayn commented:
It seems that all Minecraft versions use a damage source from 1.8 onwards to account for invulnerability. Just to satisfy my curiosity and also to get back into touch with plugin dev and nms work I decided to make a version agnostic simple kill command (basically just suicide) @ https://github.com/zp4rker/bukkitplayground/blob/kill_command/src/main/java/com/zp4rker/bukkitplayground/commands/CmdKill.java (warning: it's a little messy. Tested on Paper 1.8-1.21, haven't tested on Spigot yet but in theory should work)

I could probably implement this into a reflection provider but I don't know if I'm missing some better way to do this.

abulujayn commented:
> It seems that all Minecraft versions use a damage source from 1.8 onwards to account for invulnerability. Just to satisfy my curiosity and also to get back into touch with plugin dev and nms work I decided to make a version agnostic simple kill command (basically just suicide) @ https://github.com/zp4rker/bukkitplayground/blob/kill_command/src/main/java/com/zp4rker/bukkitplayground/commands/CmdKill.java (warning: it's a little messy. Tested on Paper 1.8-1.21, haven't tested on Spigot yet but in theory should work)
> 
> 
> 
> I could probably implement this into a reflection provider but I don't know if I'm missing some better way to do this.

@GrantGryczan Do you think this approach would solve the issue?

GrantGryczan commented:
Sorry, I'm not familiar enough with the Bukkit API to be sure. The important thing is to use a damage type present in the [`minecraft:bypasses_invulnerability`](https://minecraft.wiki/w/Damage_type_tag_(Java_Edition)#bypasses_invulnerability) damage type tag as the damage source in MC 1.20.5+. Or if it's a custom damage type, then add it to that tag.
